mailbox/store/src/main/java/org/apache/james/mailbox/store/event/EventFactory.java

```diff
@@ -34,74 +29,306 @@
+            
+            return new MailboxListener.Expunged(
+                sessionId,
+                user,
+                path,
+                mailboxId,
+                metaDataAsMap());
+        }
     }
 
-    public MailboxListener.FlagsUpdated flagsUpdated(MailboxSession session, Mailbox mailbox, List<UpdatedFlags> uflags) {
-        return new MailboxListener.FlagsUpdated(session.getSessionId(), session.getUser(), mailbox.generateAssociatedPath(), mailbox.getMailboxId(), uflags);
+    public static class MailboxAclUpdatedBuilder extends MailboxEventBuilder<MailboxAclUpdatedBuilder> {
+        private ACLDiff aclDiff;
+
+        public MailboxAclUpdatedBuilder aclDiff(ACLDiff aclDiff) {
+            this.aclDiff = aclDiff;
+            return this;
+        }
+
+        @Override
+        protected MailboxAclUpdatedBuilder backReference() {
+            return this;
+        }
+
+        public MailboxListener.MailboxACLUpdated build() {
+            Preconditions.checkState(aclDiff != null, "Field `aclDiff` is compulsory");
+            mailboxEventChecks();
+
+            return new MailboxListener.MailboxACLUpdated(
+                sessionId,
+                user,
+                path,
+                aclDiff,
+                mailboxId);
+        }
     }
 
-    public MailboxListener.MailboxRenamed mailboxRenamed(MailboxSession session, MailboxPath from, Mailbox to) {
-        return mailboxRenamed(session.getSessionId(), session.getUser(), from, to);
+    public static class MailboxAddedBuilder extends MailboxEventBuilder<MailboxAddedBuilder> {
+        @Override
+        protected MailboxAddedBuilder backReference() {
+            return this;
+        }
+
+        public MailboxListener.MailboxAdded build() {
+            mailboxEventChecks();
+
+            return new MailboxListener.MailboxAdded(sessionId, user, path, mailboxId);
+        }
+    }
+
+    public static class MailboxDeletionBuilder extends MailboxEventBuilder<MailboxDeletionBuilder> {
+        private QuotaRoot quotaRoot;
+        private QuotaCount deletedMessageCount;
+        private QuotaSize totalDeletedSize;
+
+        @Override
+        protected MailboxDeletionBuilder backReference() {
+            return this;
+        }
+
+        public MailboxDeletionBuilder quotaRoot(QuotaRoot quotaRoot) {
+            this.quotaRoot = quotaRoot;
+            return this;
+        }
+
+        public MailboxDeletionBuilder deletedMessageCount(QuotaCount deletedMessageCount) {
+            this.deletedMessageCount = deletedMessageCount;
+            return this;
+        }
+
+        public MailboxDeletionBuilder totalDeletedSize(QuotaSize totalDeletedSize) {
+            this.totalDeletedSize = totalDeletedSize;
+            return this;
+        }
+
+        public MailboxListener.MailboxDeletion build() {
+            mailboxEventChecks();
+            Preconditions.checkState(quotaRoot != null, "Field `quotaRoot` is compulsory");
```

Comment:
`Preconditions.checkNotNull`?

Reply:
Well IMO that's a state stuff: you did not set the state of your builder well which result in something being null.

If you present it like this, the state information becomes more important IMO